app/models/serial_work.rb

```diff
@@ -24,6 +24,11 @@ def delete_empty_series
     end
   end
 
+  # Reindex works added to or removed from a series
+  def update_work_index
+    work&.enqueue_to_index
```

Comment:
When is the work potentially nil? Does this differ whether the work is being deleted and the serial work follows, as opposed to the work being removed from a series (but not deleted)?

Reply:
When the work is destroyed, the associated serial works should be destroyed first, so work should be not nil in that case.

https://github.com/otwcode/otwarchive/blob/7c6681c62bab9bbfac6fc1a0ca3867e4247309b6/app/models/work.rb#L25

I'm actually not sure when or how work can be nil, but we now test that serial works can be destroyed properly even when its work is nil (#3483), so we need this nil check to avoid breaking the test.

Reply:
It can be caused by a race condition. You can simulate it in the dev console by doing the following:

1. Create a work to experiment with:
```
FactoryGirl.create(:work)
```
2. Start a transaction to add the work to a brand new series, but use `sleep` to make sure that the transaction doesn't finish quickly
```
Work.transaction do
  User.current_user = Work.last.users.first
  Work.last.update(series_attributes: { title: "Brand New Series" })
  sleep 30
end
```
3. Before that transaction finishes, open a second window with a second instance of the dev console, and destroy the work:
```
Work.last.destroy
```
4. Wait for the both transactions to finish.
5. Check `SerialWork.last.work` and note that it's `nil`.

This isn't a perfect reflection of the steps that would trigger this in practice, but it's close enough to give an idea of how it could happen: If a work is added to a series and deleted at almost exactly the same time, there's a (very low) chance that the SQL queries could be interleaved to create a `SerialWork` with a `nil` work.

Reply:
Awesome, thank you! Race conditions are some of the trickiest to wrap my head around.